src/Container/Container.php

```diff
@@ -165,4 +222,44 @@ public function __call($name, $arguments)
             throw new \BadMethodCallException("No application method named {$name}()");
         }
     }
+
+    /**
+     * Instantiate an object of named class using reflection to recursively to resolve dependencies
+     *
+     * @param string $className
+     * @return mixed
+     * @throws \ReflectionException
+     */
+    protected function resolveInstance(string $className) {
+        $class = new \ReflectionClass($className);
+
+        if (!$class->isInstantiable()) {
+            throw new \ReflectionException("$className cannot be instantiated");
+        }
+
+        return $class->newInstanceArgs($this->resolveParameters($class->getConstructor()->getParameters()));
+    }
+
+    /**
+     * Recursively resolve function parameters using reflection
```

Comment:
maybe explain the "recursively" part a little bit more to make clear that this is not a classic recursion method but recursion acutally involves several functions?

Reply:
Mutual recursion doesn't strike me as all that unusual, an the "recursively" refers more to the dependencies themselves (and their dependencies, and their dependencies, and...)